### ISSUE TITLE
Servers show "Web service is down" alert even if they don't run a web server

### ISSUE THREAD

**Issue by eburgueno** [velinath/zabbixweb#8]
Right now we don't check if the server is supposed to have Apache or IIS running, and assume they do by default. This results in servers that do not have this service running to show in alert status in the dashboard.

I thought about giving each server a checkbox 'Webserver' and store the preference, but I think it's better if we check if the trigger is enabled or not in Zabbix itself. I'll see if I can hack some code for this when I get some time.

**Comment by velinath** [comment]
I think that's actually based on a Zabbix trigger (if you enable the App
HTTP Server template) - it shouldn't return an alert for non webservers.

I can double check this when my Zabbix server comes back up

**Comment by eburgueno** [comment]
I see. It looks like there are there different scenarios we need to consider:

1) In your environment, the hosts that are not webservers don't get assigned the App HTTP Server template. All the template does is to add and item that checks for 'net.tcp.service[http]', and a trigger that alerts if the value of that item is 0 (meaning the port is not listening).
Since the item is not being linked to your other hosts, this section of the code will never run:
```
238:  case 'net.tcp.service[http]':
239:                  if($(this)[0].lastvalue == 0) {
```
This is as expected.

2) In my environment, the 'net.tcp.service[http]' is part of the default Linux Server template. This is because I've been using Zabbix since version 1.8 and that's how the template was configured back then.
The item is **enabled** for all servers, but the trigger is disabled if the server is not a webserver.
This configuration is useful if you don't care about alerts for an 'ad-hoc' or temporary webserver running, but you still want to collect the data of when it was running.

As far as I can see, the API doesn't return anything with the item that would be useful to determine if there is a trigger associated with it. We would need to change the host.get call to also include a selectTriggers so we can check for this.

3) Some sites might not use either approach. They might have the 'net.tcp.service[http]' item linked to all hosts, but disabled on the ones that don't run webservers. This would be similar to your environment, but we would need to check if the item is actually enabled before assuming we have to alert in case the lastvalue=0. If the item is not enabled, then we don't care.

```
itemid: "22785"
key_: "net.tcp.service[http]"
lastclock: "1420579542"
lastvalue: "0"
name: "WEB (HTTP) server is running"
units: ""
value_type: "3"
```

Based on the above, we can summarize the different options:

1. Item 'net.tcp.service[http]' does not exist on the host. Trigger existence or status is irrelevant.
2. Item 'net.tcp.service[http]' exists, and is enabled on the host. Trigger '{<host>:net.tcp.service[http].last(0)}=0' can be enabled or disabled.
3.  Item 'net.tcp.service[http]' exists, but disabled on the host, Trigger existence or status is irrelevant.

In pseudocode:
```
if ( item('net.tcp.service[http]').status = enabled AND 'trigger({<host>:net.tcp.service[http].last(0)}=0').status = enabled AND 'trigger({<host>:net.tcp.service[http].last(0)}=0').value = 1 ); then
  thisHostObj.error = 1;
  thisHostObj.errorText = 'Web service is down (httpd or IIS)';
end if
```

I believe this would cover all three scenarios.

I suggest the following:
a) add 'status' to the selectItems array in the host.get API call, so we can check if the item is actually enabled or not. This is to check scenarios 1 and 3.
b) add selectTriggers to the host.get API call, so we can check for the status of the trigger. This is to check scenario 2.

Sorry for the long text. I hope this makes sense. Let me know what you think!

**Comment by eburgueno** [comment]
Ok, I had a hack around it, and it turns out that even adding a 'selectTriggers' to the host.get API call we don't get anything useful.

The returned data includes a "triggers" object for the host, and the attributes are described [here](https://www.zabbix.com/documentation/2.4/manual/api/reference/trigger/object). You'd think you could associate the expression with the item key or at least the item id somehow, but all you get is a number that doesn't seem related. [This link](https://www.zabbix.com/forum/showthread.php?t=24137)  explains that the value is actually a reference to a 'function' table. The 'select_functions' query mentioned there can only be used in a triggers.get API call.

Bottom line is, I had to use something else. but I like the new logic (even if it's somewhat detrimental to performance in my environment):
1. Get all hosts and their items.
2. For each server, and each item, parse their last values.
3. If 'net.tcp.service[http]' is being checked, retrieve the trigger and it's status on that host.
4. If the trigger is enabled, and it has a value of 'problem', only then we alert.

**Comment by velinath** [comment]
That makes sense. Yeah, the triggers is not something that I got too far
into when I started working on the API - was trying to limit it to host.get
for simplicity's sake. If that's a problem when it comes to stuff like
this, I agree that it's best to sacrifice the "one API call" thing to get
the thing working across different environments.
